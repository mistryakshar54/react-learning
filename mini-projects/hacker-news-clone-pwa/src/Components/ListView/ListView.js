import React ,{ useMemo, useEffect }  from 'react';
import { makeStyles, StylesProvider } from "@material-ui/core/styles";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import Divider from "@material-ui/core/Divider";
import withLoader from "../LoadingHOC/Loader";
import Typography from "@material-ui/core/Typography";
import { Person, Schedule } from "@material-ui/icons";


const useStyles = makeStyles(theme => ({
  root: {
    width: "100%",
    backgroundColor: theme.palette.background.paper
  },
  listItemContainer: {
    display: "flex",
    flexDirection:'column',
    flex : '1 1 168px'
  },
  listItemAvatar : {
    alignItems: 'center',
    display: 'flex',
    flexDirection: 'column',
    flex : '0 1 100px',
    fontWeight : 'bold'
  },
  listItemPoints: {
    color: theme.palette.primary.main,
    fontSize : '1.7em',
  },
  listItemHeader:{
    fontSize:'1.8em',
    fontWeight:'bold'
  }
}));

const ListView = (props) => {
  const classes = useStyles();
  const {listData, currentPageNo} = props;
  return (
    <List component="nav" className={classes.root} aria-label="mailbox folders">
      {
        listData.map( (listItem , index) => {
          return (
            <div key={`news-item-${index}`}>
              <ListItem alignItems="flex-start">
                <div className={classes.listItemAvatar}>
                  <div className={classes.listItemPoints}>
                    {listItem.points}
                  </div>
                  <div>Points</div>
                </div>
                <div className={classes.listItemContainer}>
                  <Typography
                    component="div"
                    className={classes.listItemHeader}
                  >
                    {listItem.title}
                  </Typography>
                  <div style={{ display: "flex", flexDirection: "row" }}>
                    <span style={{ display: "flex" }}>
                      <Person /> &nbsp;{listItem.user}
                    </span>
                    <span style={{ display: "flex", marginLeft: "3%" }}>
                      <Schedule /> &nbsp; {listItem.time_ago}
                    </span>
                  </div>
                </div>
                {/* <ListItemText
                  primary={listItem.title}
                  secondary={
                    <React.Fragment>
                      <Typography
                        component="span"
                        variant="body2"
                        className={classes.inline}
                        color="textPrimary"
                      >
                        {listItem.user}
                      </Typography>
                      - {listItem.time_ago}
                    </React.Fragment>
                  }
                /> */}
              </ListItem>
              <Divider />
            </div>
          );
        })
      }
    </List>
  );
}
export default withLoader(ListView);

