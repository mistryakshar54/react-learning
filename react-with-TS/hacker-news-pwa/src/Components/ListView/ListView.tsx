import React from 'react';

const ListView = () => {

}

export default ListView;