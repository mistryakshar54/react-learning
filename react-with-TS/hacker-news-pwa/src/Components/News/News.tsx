import React, { useEffect, useState } from 'react';
import { RouteComponentProps, useHistory } from 'react-router-dom';
import instance from "../../AxiosConfig/Config";
import ListView from '../ListView/ListView';
import Pagination from '../Pagination/Pagination';
import { CircularProgress } from '@material-ui/core';

export type NewsType = {
    comments_count: number
    domain: string
    id: number
    points: number
    time: number
    time_ago: string
    title: string
    type: string
    url: string
    user: string
}


const News = ( { match } : RouteComponentProps ) => {

    const [appData, setData] = useState<NewsType[]>([]);
    const [appLoading, setLoading] = useState(false);
    const [appError, setError] = useState('');
   
    let [,endPoint, pageNo] = match.url.split('/');
    let history = useHistory();

    useEffect( () => {
        setLoading(true);
        const fetchNews = async() => {
            let url = parseInt(pageNo) > 0 ? endPoint + "?page=" + pageNo : endPoint;
            try {
                let newsList: NewsType[] = [];
                const newsResponse = await instance.get(url);
                if (newsResponse?.data?.length){
                    newsList = newsResponse.data;
                }
                setData(newsList);
            } catch (error) {
                if(error?.message){
                   const errorString = `${error?.response?.status} - ${error.message}`;
                   setError(errorString);
                }
                else{
                    setError('An error has occuerd!!');
                }
            }
            setLoading(false);        
        }
        fetchNews();
    }, [endPoint, pageNo]);

    const LoadingContent = () => <div style={ { display : 'flex', justifyContent : 'center', margin:'10%' } }><CircularProgress color="primary" /></div>;
    const ErrorContent = () => <h1>Error! {appError}</h1>;

    const handleNavigatePage = (direction : string) => {
        let currPageNo = parseInt(pageNo); 
        direction === 'previous' ? currPageNo-- : currPageNo++;
        history.push(`/${endPoint}/${currPageNo}`);
    }

    return (
        <div>
            {
                appLoading ? <LoadingContent/> : 
                appError ? <ErrorContent/> :
                <div>
                    <Pagination currentPage={parseInt(pageNo)} nextPage={handleNavigatePage} />
                    <ListView newsData={ appData } />
                </div>
            }
        </div>
    );
}

export default News;