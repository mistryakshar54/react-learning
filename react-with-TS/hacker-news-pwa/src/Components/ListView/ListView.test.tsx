import React from "react";
import Enzyme, { mount } from 'enzyme';
import ListView, { ListViewProps, styles } from "./ListView";
import Adapter from "enzyme-adapter-react-16";
import renderer from "react-test-renderer";
import { NewsType } from "../News/News";

Enzyme.configure({ adapter: new Adapter() });

const defaultProps = {
    newsData: [
        {
            comments_count: 1,
            domain: 'abc',
            id: 1,
            points: 1,
            time: 123,
            time_ago: 'abc',
            title: 'sample',
            type: 'string',
            url: 'myUrl',
            user: 'ABC'
        },
        {
            comments_count: 3,
            domain: 'abc',
            id: 13,
            points: 1,
            time: 123,
            time_ago: 'abc',
            title: 'sample',
            type: 'string',
            url: 'myUrl',
            user: 'ABC'
        }
    ]
}

const createShallowComponent = (props = defaultProps) => {
    return mount(<ListView {...props} />)
}

test("renders snapshot", () => {
    const component = renderer.create(
        <ListView { ...defaultProps } />
    );
    expect(component).toMatchSnapshot();
});

test('Shows empty when there is no data', () => {
    const wrapper = createShallowComponent({ newsData : [] });
    const emptyHeader = wrapper.find('#emptyListHeader');
    expect(emptyHeader.length).toBe(1);
    expect(emptyHeader.text()).toBe('No Data!!')
});

test('Shows list when there is data', () => {
    const wrapper = createShallowComponent();
    const listItems = wrapper.findWhere( node => node.name() === 'li'   )
    expect(listItems.length).toBe(2);
});

test('Shows list when there is data', () => {
    const wrapper = createShallowComponent();
    const listItems = wrapper.findWhere( node => node.name() === 'li'   )
    expect(listItems.length).toBe(2);
});
