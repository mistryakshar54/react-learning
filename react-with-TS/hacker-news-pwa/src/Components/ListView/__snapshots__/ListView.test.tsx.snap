// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation3 MuiPaper-rounded"
  style={
    Object {
      "marginTop": "1%",
    }
  }
>
  <nav
    aria-label="mailbox folders"
    className="MuiList-root ListView-root-1 MuiList-padding"
  >
    <div>
      <li
        className="MuiListItem-root list-item MuiListItem-gutters MuiListItem-alignItemsFlexStart"
        disabled={false}
      >
        <div
          className="ListView-listItemAvatar-3"
        >
          <div
            className="ListView-listItemPoints-4"
          >
            1
          </div>
          <div>
            Points
          </div>
        </div>
        <div
          className="ListView-listItemContainer-2"
        >
          <div
            className="MuiTypography-root MuiTypography-body1"
          >
            <a
              className="ListView-listItemHeader-5"
              href="myUrl"
              target="_blank"
            >
              sample
            </a>
          </div>
          <div
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
              }
            }
          >
            <span
              className="ListView-newsMetaData-6"
              style={
                Object {
                  "marginRight": "10px",
                }
              }
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root ListView-icon-7"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 6c1.1 0 2 .9 2 2s-.9 2-2 2-2-.9-2-2 .9-2 2-2m0 10c2.7 0 5.8 1.29 6 2H6c.23-.72 3.31-2 6-2m0-12C9.79 4 8 5.79 8 8s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4zm0 10c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                />
              </svg>
                
              ABC
            </span>
            <span
              className="ListView-newsMetaData-6"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root ListView-icon-7"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
                <path
                  d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z"
                />
              </svg>
                
              abc
            </span>
          </div>
        </div>
      </li>
      <hr
        className="MuiDivider-root"
      />
    </div>
    <div>
      <li
        className="MuiListItem-root list-item MuiListItem-gutters MuiListItem-alignItemsFlexStart"
        disabled={false}
      >
        <div
          className="ListView-listItemAvatar-3"
        >
          <div
            className="ListView-listItemPoints-4"
          >
            1
          </div>
          <div>
            Points
          </div>
        </div>
        <div
          className="ListView-listItemContainer-2"
        >
          <div
            className="MuiTypography-root MuiTypography-body1"
          >
            <a
              className="ListView-listItemHeader-5"
              href="myUrl"
              target="_blank"
            >
              sample
            </a>
          </div>
          <div
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
              }
            }
          >
            <span
              className="ListView-newsMetaData-6"
              style={
                Object {
                  "marginRight": "10px",
                }
              }
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root ListView-icon-7"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 6c1.1 0 2 .9 2 2s-.9 2-2 2-2-.9-2-2 .9-2 2-2m0 10c2.7 0 5.8 1.29 6 2H6c.23-.72 3.31-2 6-2m0-12C9.79 4 8 5.79 8 8s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4zm0 10c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                />
              </svg>
                
              ABC
            </span>
            <span
              className="ListView-newsMetaData-6"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root ListView-icon-7"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
                <path
                  d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z"
                />
              </svg>
                
              abc
            </span>
          </div>
        </div>
      </li>
      <hr
        className="MuiDivider-root"
      />
    </div>
  </nav>
</div>
`;
